python/lsst/ip/isr/calibType.py

```diff
@@ -0,0 +1,600 @@
+
+        Parameters
+        ----------
+        setDate : `bool, optional
+            Update the CALIBDATE fields in the metadata to the current
+            time. Defaults to False.
+        kwargs :
+            Other keyword parameters to set in the metadata.
+        """
+        kwargs["DETECTOR"] = self._detectorName
+        kwargs["DETECTOR_SERIAL"] = self._detectorSerial
+
+        kwargs['INSTRUME'] = self.instrument
+        kwargs['calibType'] = self.calibType
+        super().updateMetadata(setDate=setDate, **kwargs)
+
+    def fromDataIds(self, dataIdList):
+        """Update provenance from dataId List.
+
+        Parameters
+        ----------
+        dataIdList : `list` [`lsst.daf.butler.DataId`]
+            List of dataIds used in generating this calibration.
+        """
+        for dataId in dataIdList:
+            for key in dataId:
+                if key not in self.dimensions:
+                    self.dimensions.add(key)
+            self.dataIdList.append(dataId)
+
+    @classmethod
+    def fromTable(cls, tableList):
+        """Construct provenance from table list.
+
+        Parameters
+        ----------
+        tableList : `list` [`lsst.afw.table.Table`]
+            List of tables to construct the provenance from.
+
+        Returns
+        -------
+        provenance : `lsst.ip.isr.IsrProvenance`
+            The provenance defined in the tables.
+        """
+        table = tableList[0]
+        metadata = table.meta
+        inDict = dict()
+        inDict['metadata'] = metadata
+        inDict['detectorName'] = metadata['DETECTOR']
+        inDict['detectorSerial'] = metadata['DETECTOR_SERIAL']
+        inDict['instrument'] = metadata['INSTRUME']
+        inDict['calibType'] = metadata['calibType']
+        inDict['dimensions'] = set()
+        inDict['dataIdList'] = list()
+
+        schema = dict()
+        for colName in table.columns:
+            schema[colName.lower()] = colName
+            inDict['dimensions'].add(colName.lower())
+        inDict['dimensions'] = sorted(inDict['dimensions'])
+
+        for row in table:
+            entry = dict()
+            for dim in sorted(inDict['dimensions']):
+                entry[dim] = row[schema[dim]]
+            inDict['dataIdList'].append(entry)
+
+        return cls.fromDict(inDict)
+
+    @classmethod
+    def fromDict(cls, dictionary):
+        """Construct provenance from a dictionary.
+
+        Parameters
+        ----------
+        dictionary : `dict`
+            Dictionary of provenance parameters.
+
+        Returns
+        -------
+        provenance : `lsst.ip.isr.IsrProvenance`
+            The provenance defined in the tables.
+        """
+        calib = cls()
+        calib.updateMetadata(setDate=False, **dictionary['metadata'])
+        calib._detectorName = dictionary['detectorName']
+        calib._detectorSerial = dictionary['detectorSerial']
+        calib.instrument = dictionary['metadata']['INSTRUME']
```

Comment:
In fromTable doesn't an "instrument" field get filled in to the dict so I'm not sure why we need to look in the metadata here.